mat64/dense_arithmetic.go

```diff
@@ -30,17 +30,17 @@ func (m *Dense) Add(a, b Matrix) {
 	bMat, _ := untranspose(b)
 	m.reuseAs(ar, ac)
 	var restore func()
-	if m == aMat {
-		m, restore = m.isolatedWorkspace(aMat)
-		defer restore()
-	} else if m == bMat {
-		m, restore = m.isolatedWorkspace(bMat)
-		defer restore()
-	}
 
-	if a, ok := a.(RawMatrixer); ok {
-		if b, ok := b.(RawMatrixer); ok {
-			amat, bmat := a.RawMatrix(), b.RawMatrix()
+	if arm, ok := a.(RawMatrixer); ok {
+		if brm, ok := b.(RawMatrixer); ok {
+			amat, bmat := arm.RawMatrix(), brm.RawMatrix()
```

Comment:
The idiom I've used here is explained in https://github.com/gonum/matrix/pull/263/files#diff-928774397024853172d191bc227c4939R23. It is somewhat misleading and I apologise for that; if given enough of an argument I would change it, but I would rather not.

In short, if checkOverlap does not return false, it panics, so the check is not one that will result in an isolation. The reason for doing it this way is that it keeps the check and the pointer comparison next to each other, which I think balances out the C-ism.